pkg/ebpf/tracee.go

```diff
@@ -942,49 +994,28 @@ func (t *Tracee) populateBPFMaps() error {
-	binary.LittleEndian.PutUint32(configVal[8:12], t.getFiltersConfig())
-	binary.LittleEndian.PutUint32(configVal[12:16], uint32(t.containers.GetDefaultCgroupHierarchyID()))
-	binary.LittleEndian.PutUint64(configVal[16:24], t.config.Filter.UIDFilter.Maximum())
-	binary.LittleEndian.PutUint64(configVal[24:32], t.config.Filter.UIDFilter.Minimum())
-	binary.LittleEndian.PutUint64(configVal[32:40], t.config.Filter.PIDFilter.Maximum())
-	binary.LittleEndian.PutUint64(configVal[40:48], t.config.Filter.PIDFilter.Minimum())
-	binary.LittleEndian.PutUint64(configVal[48:56], t.config.Filter.MntNSFilter.Maximum())
-	binary.LittleEndian.PutUint64(configVal[56:64], t.config.Filter.MntNSFilter.Minimum())
-	binary.LittleEndian.PutUint64(configVal[64:72], t.config.Filter.PidNSFilter.Maximum())
-	binary.LittleEndian.PutUint64(configVal[72:80], t.config.Filter.PidNSFilter.Minimum())
-	// Next 128 bytes (1024 bits) are used for events_to_submit configuration
-	// Set according to events chosen by the user
-	for id, e := range t.events {
-		if id >= 1024 {
-			// we support up to 1024 events shared with bpf code
-			continue
-		}
-		if e.submit {
-			index := id / 8
-			offset := id % 8
-			configVal[80+index] = configVal[80+index] | (1 << offset)
-		}
-	}
+	configVal := t.computeConfigValues()
 	if err = bpfConfigMap.Update(unsafe.Pointer(&cZero), unsafe.Pointer(&configVal[0])); err != nil {
 		return err
 	}
 
-	errmap := make(map[string]error, 0)
-	errmap[UIDFilterMap] = t.config.Filter.UIDFilter.InitBPF(t.bpfModule)
-	errmap[PIDFilterMap] = t.config.Filter.PIDFilter.InitBPF(t.bpfModule)
-	errmap[MntNSFilterMap] = t.config.Filter.MntNSFilter.InitBPF(t.bpfModule)
-	errmap[PidNSFilterMap] = t.config.Filter.PidNSFilter.InitBPF(t.bpfModule)
-	errmap[UTSFilterMap] = t.config.Filter.UTSFilter.InitBPF(t.bpfModule)
-	errmap[CommFilterMap] = t.config.Filter.CommFilter.InitBPF(t.bpfModule)
-	errmap[ContIdFilter] = t.config.Filter.ContIDFilter.InitBPF(t.bpfModule, t.containers)
-	errmap[BinaryFilterMap] = t.config.Filter.BinaryFilter.InitBPF(t.bpfModule)
+	for filterScope := range t.config.FilterScopes.Map() {
+		scopeID := uint(filterScope.ID)
+		errMap := make(map[string]error, 0)
 
-	for k, v := range errmap {
-		if v != nil {
-			return fmt.Errorf("error setting %v filter: %v", k, v)
+		errMap[UIDFilterMap] = filterScope.UIDFilter.UpdateBPF(t.bpfModule, scopeID)
+		errMap[PIDFilterMap] = filterScope.PIDFilter.UpdateBPF(t.bpfModule, scopeID)
+		errMap[MntNSFilterMap] = filterScope.MntNSFilter.UpdateBPF(t.bpfModule, scopeID)
+		errMap[PidNSFilterMap] = filterScope.PidNSFilter.UpdateBPF(t.bpfModule, scopeID)
+		errMap[UTSFilterMap] = filterScope.UTSFilter.UpdateBPF(t.bpfModule, scopeID)
+		errMap[CommFilterMap] = filterScope.CommFilter.UpdateBPF(t.bpfModule, scopeID)
+		errMap[ContIdFilter] = filterScope.ContIDFilter.UpdateBPF(t.bpfModule, t.containers, scopeID)
+		errMap[BinaryFilterMap] = filterScope.BinaryFilter.UpdateBPF(t.bpfModule, scopeID)
```

Comment:
I still see that you loop over all scopes and update the bpf map for every scope.
The value to be update can be calculated once, without the need to get the value from the bpf map, modify, and the update the bpf map again for each scope.

Reply:
> The value to be update can be calculated once, without the need to get the value from the bpf map, modify, and the update the bpf map again for each scope.

Unfortunately it can't, at least with the current design, if I'm not wrong.

Each filterScope has a bunch of filters like UIDFilterMap, PIDFilterMap, CommFilterMap etc. And each one of those can have different keys and values.

The cmdline:

`-t 2:comm=shell.sh -t 4:comm=script.sh,zsh`

builds two scope each with its own set of filters, but with only comm filters enabled and set.

Each iteration works on updating (initializing) the comm map for the respective scope, see:

![image](https://user-images.githubusercontent.com/3372117/213571892-c1c0e946-39c8-4cf9-b1db-72bd096009d3.png)

![image](https://user-images.githubusercontent.com/3372117/213572148-c179ea68-299e-49ad-9546-1f8c87ab4c3a.png)

So the rest of the work in the UpdateBPF() is to set or clear the scope bit, based on the equality precedence, which must be done for each equality/non-equality. In our example:

comm=shell.sh in the scope 2
comm=script.sh and comm=zsh in the scope 4

Which results in the comm_filter map populated as:

![image](https://user-images.githubusercontent.com/3372117/213572934-21fd88cf-77c5-4559-b1b2-4f49e4ea2505.png)

To avoid iterating through the filterScope, we must redesign where we keep information about scopes. For now I see no other practical way.

Reply:
> To avoid iterating through the filterScope...

This is not what I meant. We do need to iterate it. What I meant is that we should do the calculation not using the bpf map itself, but using user-defined maps. Get, modify, update can be done on these user maps, and then update the bpf map only once

Reply:
After discussion we've decided that it can be tackled in a future PR.